pkg/retry/retry.go

```diff
@@ -0,0 +1,112 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package retry
+
+import (
+	"math/rand"
+	"time"
+)
+
+// RetryableFunc represents a function that can be retried.
+type RetryableFunc func() error
+
+// Retryer defines the requirements for retrying a function.
+type Retryer interface {
+	Retry(RetryableFunc) error
+}
+
+// Ticker defines the requirements for providing a clock to the retry logic.
+type Ticker interface {
+	Tick() time.Duration
+	Stop() <-chan struct{}
+}
+
+// TimeoutError represents a timeout error.
+type TimeoutError struct{}
+
+func (TimeoutError) Error() string {
+	return "timeout"
+}
+
+// IsTimeout reutrns if the provided error is a timeout error.
+func IsTimeout(err error) bool {
+	_, ok := err.(TimeoutError)
+
+	return ok
+}
+
+type expectedError struct{ error }
+
+type unexpectedError struct{ error }
+
+type retryer struct {
+	duration time.Duration
+	options  *Options
+}
+
+type ticker struct {
+	C       chan time.Time
+	options *Options
+	rand    *rand.Rand
+	s       chan struct{}
+}
+
+func (t ticker) Jitter() time.Duration {
+	if int(t.options.Jitter) == 0 {
+		return time.Duration(0)
+	}
+
+	if t.rand == nil {
+		t.rand = rand.New(rand.NewSource(time.Now().UnixNano()))
+	}
+
+	return time.Duration(t.rand.Int63n(int64(t.options.Jitter)))
+}
+
+func (t ticker) Stop() <-chan struct{} {
+	return t.s
+}
+
+func (t ticker) stop() {
+	t.s <- struct{}{}
+}
+
+// ExpectedError error represents an error that is expected by the retrying
+// function. This error is ignored.
+func ExpectedError(err error) error {
+	return expectedError{err}
+}
+
+// UnexpectedError error represents an error that is unexpected by the retrying
+// function. This error is fatal.
+func UnexpectedError(err error) error {
+	return unexpectedError{err}
+}
+
+func retry(f RetryableFunc, d time.Duration, t Ticker) error {
+	timer := time.NewTimer(d)
+	defer timer.Stop()
+
+	for {
+		select {
+		case <-timer.C:
+			return TimeoutError{}
+		case <-t.Stop():
+			return nil
+		default:
+			if err := f(); err != nil {
+				switch err.(type) {
+				case expectedError:
+					time.Sleep(t.Tick())
```

Comment:
@smira I think this does what we discussed in slack? So the total time taken is now `time(f()) + time.Sleep(t.Tick())`.

Reply:
if we need it to be stoppable while sleeping, it makes sense to rewrite this a bit:

```suggestion
					select {
					case <-time.After(t.Tick()):
					case <-t.Stop():
					    return nil
					 }
```

Reply:
and I don't know how to actually handle all the cases if we want to:

* global timeout exceeded
* stop
* wait between calls
* in a call to `f()`

If we want it to be stoppable all the time, probably we should have a context which cancels both `f()` and any wait we have here, so on the surface this might be:

```
1. call f(ctx)
2. wait for either: ctx cancel, end of interval between retries, global timeout
```

if we want global timeout to cancel `f(ctx)` while running, we need to run `f` in goroutine and cancel its `ctx` if global timeout is reached...

so much here but not sure what kind of requirements we have

Reply:
For now our needs are pretty basic. We just need to attempt something for a given duration. Fair to say we can follow up with a more complete implementation?